servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,705 @@
+                final int prevActiveMappedSources = activeMappedSources;
+                if (prevActiveMappedSources < 0) {
+                    // We have been cancelled, or already completed and the active count flipped to negative, either way
+                    // we don't want to Subscribe or retain a reference to this Publisher.
+                    subscribers.remove(subscriber);
+                    break;
+                } else if (activeMappedSourcesUpdater.compareAndSet(this, prevActiveMappedSources,
+                        prevActiveMappedSources + 1)) {
+                    publisher.subscribeInternal(subscriber);
+                    tryRequestMore(subscriber);
+                    break;
+                }
+            }
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            if (!onError0(t, false, false)) {
+                LOGGER.debug("Already terminated/cancelled, ignoring error notification.", t);
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            // Setting terminal must be done before terminateActiveMappedSources to ensure visibility of the terminal.
+            final boolean setTerminal = trySetTerminal(complete(), false, terminalNotificationUpdater, this);
+            final boolean allSourcesTerminated = terminateActiveMappedSources();
+            if (setTerminal && allSourcesTerminated) {
+                enqueueAndDrain(complete());
+            }
+        }
+
+        private boolean onError0(Throwable throwable, boolean overrideComplete,
+                                 boolean cancelSubscriberIfNecessary) {
+            final TerminalNotification notification = TerminalNotification.error(throwable);
+            if (trySetTerminal(notification, overrideComplete, terminalNotificationUpdater, this)) {
+                try {
+                    doCancel(cancelSubscriberIfNecessary);
+                } finally {
+                    enqueueAndDrain(notification);
+                }
+                return true;
+            }
+            return false;
+        }
+
+        private void doCancel(boolean cancelSubscription) {
+            Throwable delayedCause = null;
+            // Prevent future onNext operations from adding to subscribers which otherwise may result in
+            // not cancelling mapped Subscriptions. This should be Integer.MIN_VALUE to prevent subsequent mapped
+            // source completion from incrementing the count to 0 or positive as terminateActiveMappedSources flips
+            // the count to negative to signal to mapped sources we have completed.
+            activeMappedSources = Integer.MIN_VALUE;
+            try {
+                if (cancelSubscription) {
+                    assert subscription != null;
+                    subscription.cancel();
+                }
+            } finally {
+                for (FlatMapPublisherSubscriber<T, R> subscriber : subscribers) {
```

Comment:
is it possible that `cancel()` and `onNext(...)` is called concurrently ? If so just iterating over the subscribers and then call `clear()` is racy as you may miss some subscribers that were added by `onNext(...)`

Reply:
Yes this is possible. I believe this case is accounted for as follows:

- `cancel()` sets `activeMappedSources = Integer.MIN_VALUE;` before iterating/clearing the map. See comments from `onCancel(..)`:
```
// Prevent future onNext operations from adding to subscribers which otherwise may result in
// not cancelling mapped Subscriptions. This should be Integer.MIN_VALUE to prevent subsequent mapped
// source completion from incrementing the count to 0 or positive as terminateActiveMappedSources flips
// the count to negative to signal to mapped sources we have completed.
```

- `onNext(..)` attempts to atomically increment `activeMappedSources` after adding to the map, and removes the entry if failed to increment because negative value is seen (negative values are also used when the `FlatMapSubscriber` terminates to coordinate with `FlatMapPublisherSubscriber` completion). See comments from `onNext(..)`:
```
// We have been cancelled, or already completed and the active count flipped to negative, either way
// we don't want to Subscribe or retain a reference to this Publisher.
```